### PR TITLE
CR-1133835 Print progress report when flashing SC using xgq's

### PR DESCRIPTION
**Pull request by rbramand-xilinx**
#### Problem solved by the commit
When flashing SC using xgq's on vck5000 kind of devices continuous dots will be printed to indicate progress

#### How problem was solved, alternative solutions (if any) and why they were rejected
Flashing SC is made as blocking call and while we wait progress is reported by printing continuous dots. Percentage is not printed at present as we dont have any info from VMR on how much progress is being made.

#### Risks (if any) associated the changes in the commit
none

#### What has been tested and how, request additional testing if necessary
Tested by flashing on vck5000 device and the result is as expected.

eg: time ./build/Release/opt/xilinx/xrt/bin/unwrapped/xbmgmt2 program --base all --image xilinx_vck5000_gen4x8_qdma_base_1 -d --force

Device : [0000:d8:00.0]
Current Configuration
  Platform              : xilinx_vck5000_gen4x8_qdma_base_1
  SC Version           : 4.4.33
  Platform ID          : 0xc727c8224d876f34
Incoming Configuration
  Deployment File          : partition.xsabin
  Deployment Directory : /lib/firmware/xilinx/c727c8224d876f34d05896abf1aaff02
  Size                              : 7,646,083 bytes
  Timestamp                   : Wed Jun 29 13:47:23 2022
  Platform                       : xilinx_vck5000_gen4x8_qdma_base_1
  SC Version                    : 4.4.33
  Platform UUID        : C727C822-4D87-6F34-D058-96ABF1AAFF02

Actions to perform:
  [0000:d8:00.0] : Program base (FLASH) image
  [0000:d8:00.0] : Program Satellite Controller (SC) image

Are you sure you wish to proceed? [Y/n]: Y (Force override)

INFO: Forcing flashing of the Satellite Controller (SC) image (Force flag is set).
[0000:d8:00.0] : Updating Satellite Controller (SC) firmware flash image
**..........................................................................................**

INFO: Forcing flashing of the base (e.g., shell) image (Force flag is set).
#### Documentation impact (if any)
NA

### CODE REVIEW

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 1]
```diff
@@ -683,6 +683,50 @@ struct accel_deadlock_status
...
+
+    return value;
+  }
+
+  static void
+  put(const xrt_core::device* device, key_type key, const boost::any& val)
+  {
+    auto pdev = get_pcidev(device);
+    //check for exsistence of sysfs node
+    std::string sysfs_path = pdev->get_sysfs_path(subdev, entry);
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:716]: is this function returns empty if path doesn't exist?

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 2]
```diff
@@ -683,6 +683,50 @@ struct accel_deadlock_status
...
+    }
+
+    std::string err;
+    pdev->sysfs_put(subdev, entry, err, boost::any_cast<value_type>(val));
+    if (!err.empty())
+      throw xrt_core::query::sysfs_error(err);
+  }
+};
+const std::string program_sc::subdev = "xgq_vmr";
+const std::string program_sc::entry = "program_sc";
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:728]: Why are these variables defined outside query? Please move it inside query structure.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 3]
```diff
@@ -151,10 +161,20 @@ update_SC(unsigned int  index, const std::string& file)
...
   try {
     uint32_t val = xrt_core::query::program_sc::value_type(1);
+    bool done = false;
+    // Print progress while sc is flashed
+    std::thread t(progressReporter, std::ref(done));
     xrt_core::device_update<xrt_core::query::program_sc>(dev.get(), val);
+    done = true;
+    t.join();
+    std::cout << std::endl;
+	std::cout << boost::format("%-8s : %s \n\n") % "INFO" % "SC firmware image has been programmed successfully.";
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:171]: indentation fix. why don't you merge both std::couts?

**Review by chvamshi-xilinx — CHANGES_REQUESTED** [review]
Looks good overall. Some small suggestions.

File: src/runtime_src/core/pcie/linux/device_linux.cpp (hunk 1 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/pcie/linux/device_linux.cpp:716]: Yes @chvamshi-xilinx, it returns empty string if path doesnot exist

File: src/runtime_src/core/pcie/linux/device_linux.cpp (hunk 2 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/pcie/linux/device_linux.cpp:728]: as these are static variables I have defined them outside

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp (hunk 3 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:171]: sure @chvamshi-xilinx

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 4]
```diff
@@ -151,10 +161,19 @@ update_SC(unsigned int  index, const std::string& file)
...
+    // Print progress while sc is flashed
+    std::thread t(progressReporter, std::ref(done));
     xrt_core::device_update<xrt_core::query::program_sc>(dev.get(), val);
+    done = true;
+    t.join();
+    std::cout << boost::format("\n%-8s : %s \n\n") % "INFO" % "SC firmware image has been programmed successfully.";
     return;
   }
-  catch (const xrt_core::query::exception&) {
+  catch (const xrt_core::query::sysfs_error &e) {
```
Comment (stsoe) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:173]:
```suggestion
  catch (const xrt_core::query::sysfs_error& e) {
```

**Review by stsoe — APPROVED** [review]
Looks good to me.

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 5]
```diff
@@ -137,6 +137,16 @@ is_SC_fixed(unsigned int index)
   }
 }
 
+// Function to print progress using dots
+static void
+progressReporter(bool& done)
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:142]: Are you aware of the core/tools/common/ProgressBar.cpp functionality.   

This class will perform this operation for you.

**Review by rozumx — CHANGES_REQUESTED** [review]
Please review the ProgressBar.cpp class and see if it can be integrated into the code base, instead of creating a new one.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp (hunk 5 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:142]: Hi @rozumx, yes I have checked ProgressBar.cpp file initially and tried integrating this code but there we need to feed status of the task every time using update function but at present we don't get any progress status from VMR it just returns success or failure. So created a new static function to print dots until success or failure status is returned.

If we need to use ProgressBar.cpp code one way I can think of is to set max iterations as value greater than time out from VMR and call update function every time after some fixed duration of time.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 6]
```diff
@@ -137,6 +137,17 @@ is_SC_fixed(unsigned int index)
...
+{
+  while (!done) {
+    std::cout << "." << std::flush;
+    std::this_thread::sleep_for(std::chrono::seconds(1));
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:146]: @keryell yeah but I am using seconds only once so using this way

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 7]
```diff
@@ -151,12 +162,23 @@ update_SC(unsigned int  index, const std::string& file)
   //versal flow to flash sc
   try {
     uint32_t val = xrt_core::query::program_sc::value_type(1);
+    bool done = false;
+    // Print progress while sc is flashed
+    std::thread t(progressReporter, std::ref(done));
```
[the remark this answers is not in the document]
Comment (rbramand-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp]: @keryell thanks for the suggestion, I have changed the code accordingly

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 8]
```diff
@@ -149,14 +149,35 @@ update_SC(unsigned int  index, const std::string& file)
   auto dev = xrt_core::get_mgmtpf_device(index);
 
   //versal flow to flash sc
+  std::thread t;
+  bool done;
```
Comment (Ralender) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:153]: this bool needs to be atomic there since it is accessed from different threads.

**Comment by rbramand-xilinx** [comment]
@rozumx , @chvamshi-xilinx I made changes as we discussed in meeting. Please review

File: src/runtime_src/core/tools/common/ProgressBar.h [hunk 9]
```diff
@@ -55,6 +55,7 @@ class ProgressBar {
   formatTime(std::chrono::duration<double> duration);
 
   unsigned int getMaxIterations() {return m_maxNumIterations;}
+  void setPrintPercentBatch(bool val) {m_printPercentBatch = val;};
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ProgressBar.h:58]: Give a space after { for better readability.

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 10]
```diff
@@ -137,6 +138,40 @@ is_SC_fixed(unsigned int index)
...
 
+//versal flow to flash sc
+static void
+update_versal_SC(std::shared_ptr<xrt_core::device> dev)
+{
+  std::thread t;
+  std::atomic<bool> done(false);
+  std::shared_ptr<XBU::ProgressBar> progress_reporter;
+  try {
+    uint32_t val = xrt_core::query::program_sc::value_type(1);
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:149]: use const

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 11]
```diff
@@ -137,6 +138,40 @@ is_SC_fixed(unsigned int index)
...
+    // timeout for xgq is 300 seconds
+    unsigned int max_duration = 300;
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:151]: const

File: src/runtime_src/core/tools/common/ProgressBar.cpp [hunk 12]
```diff
@@ -82,6 +82,7 @@ ProgressBar::ProgressBar(const std::string &_opNname, unsigned int _maxNumIterat
     , m_maxNumIterations(_maxNumIterations)
     , m_isBatch(_isBatch)
     , m_ostr(_ostr)
+    , m_printPercentBatch(true)
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/common/ProgressBar.cpp:85]: Change copyright at line 2

File: src/runtime_src/core/tools/common/ProgressBar.h (hunk 9 above)
Comment (chvamshi-xilinx) [reply src/runtime_src/core/tools/common/ProgressBar.h:58]: Change the copyright at line 2

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp [hunk 13]
```diff
@@ -148,16 +183,11 @@ update_SC(unsigned int  index, const std::string& file)
 
   auto dev = xrt_core::get_mgmtpf_device(index);
 
-  //versal flow to flash sc
-  try {
-    uint32_t val = xrt_core::query::program_sc::value_type(1);
-    xrt_core::device_update<xrt_core::query::program_sc>(dev.get(), val);
+  bool is_versal = xrt_core::device_query<xrt_core::query::is_versal>(dev);
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:186]: This Query is not defined in Windows. It throws an exception for Alveo devices as well. 
You need to define and return false as Windows doesnt support versal yet.

**Review by chvamshi-xilinx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp (hunk 13 above)
Comment (rbramand-xilinx) [reply src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp:186]: Added query in windows as well, thanks @chvamshi-xilinx

**Review by rozumx — APPROVED** [review]
Thank you for making the changes.

**Pull request merged**